### ISSUE TITLE
Frame.from_sql converts datetime index columns to object when index_depth is nonzero

### ISSUE THREAD

**Issue by brandtbucher** [InvestmentSystems/static-frame#326]
**Description**

It appears that building an `IndexHierarchy` in `Frame.from_sql` converts `datetime64` columns to `datetime.datetime` (`object`).

**Platform**

```py
>>> import static_frame as sf
>>> sf.Platform.display()
<Series: platform>
<Index>
platform           Linux-5.4.0-66-generic-x86_64-with-glibc2.27
sys.version        3.8.3 (default, Mar 10 2021, 08:40:47) [GCC 7.5.0]
static-frame       0.8.2
numpy              1.17.4
pandas             0.24.2
xlsxwriter         1.1.2
openpyxl           2.5.4
xarray             <ModuleNotFoundError>
tables             3.6.1
pyarrow            0.17.0
msgpack            <ModuleNotFoundError>
msgpack_numpy      <ModuleNotFoundError>
<<U13>             <object>
```

**Example**

Here is some example data, loaded from a SQL query. Note that `index_depth` is not set, and the first column of the resulting data is type `datetime64[D]` (as specified by `dtypes`):

```py
>>> sf.Frame.from_sql( 
...     QUERY, 
...     connection=connection, 
...     # index_depth=2, 
...     dtypes={ 
...         'data_date': 'datetime64[D]', 
...         'spam': 'int64', 
...         'eggs': 'float64', 
...     }, 
... ).head()
<Frame>
<Index> data_date       spam     eggs      <<U10>
<Index>
0       2020-01-01      1        42.0
1       2020-01-02      2        nan
2       2020-01-03      4        nan
3       2020-01-04      8        nan
4       2020-01-05      16       42.0
<int64> <datetime64[D]> <int64>  <float64>
```

However, making this column part of an `IndexHierarchy` (by setting `index_depth=2`) converts it to an `object` array of `datetime.datetime` objects:

```py
>>> sf.Frame.from_sql(
...     QUERY,
...     connection=connection,
...     index_depth=2,
...     dtypes={
...         'data_date': 'datetime64[D]',
...         'spam': 'int64',
...         'eggs': 'float64',
...     },
... ).head()
<Frame>
<Index>                     eggs      <<U5>
<IndexHierarchy>
2020-01-01 00:00:00 1       42.0
2020-01-02 00:00:00 2       nan
2020-01-03 00:00:00 4       nan
2020-01-04 00:00:00 8       nan
2020-01-05 00:00:00 16      42.0
<object>            <int64> <float64>
```

Here is what I would expect. Doing this manually (as shown here) makes the whole `Frame` construction take about twice as long:

```py
>>> sf.Frame.from_sql(
...     QUERY,
...     connection=connection, 
...     # index_depth=2, 
...     dtypes={
...         'data_date': 'datetime64[D]',
...         'spam': 'int64',
...         'eggs': 'float64',
...     },
... ).set_index_hierarchy(['data_date', 'spam'], drop=True).head()
<Frame>
<Index>                                      eggs      <<U5>
<IndexHierarchy: ('data_date', '...
2020-01-01                           1       42.0
2020-01-02                           2       nan
2020-01-03                           4       nan
2020-01-04                           8       nan
2020-01-05                           16      42.0
<datetime64[D]>                      <int64> <float64>

```

**Comment by flexatone** [comment]
Thanks for identifying this issue.

What is happening here is not that the values in the `IH` are being converted to `datetime.datetime`, the values are simply not being converted at all, i.e., these are the objects that `pyodbc` is giving you from the query. The reason that they are not being converted is that `Frame.from_sql` uses `Frame.from_records`; in `Frame.from_records`, the `dtypes` argument specifies dtypes only for true columns: the `Index` values are taken as is without conversion. As `Frame.from_records` does not expose an `index_depth` argument, this makes sense in that context, but is obviously a shortcoming in the context of `Frame.from_sql`.

I will give some thought to how this might be fixed.

**Comment by flexatone** [comment]
This is fixed and will be part of the 0.8.5 release. 

The answer was to make `Frame.from_sql` more sophisticated in how it prepares the `index` and `index_constructor` arguments given to `Frame.from_records`.

**Comment by brandtbucher** [comment]
Thanks for the quick fix!
